### PR TITLE
TASK: Add support for ElasticSearch 2.*

### PR DESCRIPTION
**Pull request by dfeyer**
This change add support for recent version of ElasticSearch. This change is not backward compatible and break the support for earlier version.

### CODE REVIEW

**Comment by dfeyer** [comment]
This change is Work In Progress

- [x] Adapt change in inline script API
- [x] Remove the delete query
- [x] Implement BulkIndexerInterface to disable some parts of the logic when building the full index (performance)
- [x] Skip unstructured node type from Indexing (break ES 2.x more strict index configuration)
- [ ] Replace the delete query by a scroll API + bulk delete

**Comment by ComiR** [comment]
see https://github.com/ElementareTeilchen/Flowpack.ElasticSearch.ContentRepositoryAdaptor/commit/c3936495accacae0bcb3ab82b5a516e54bf4bde4 for the last one.

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php [hunk 1]
```diff
@@ -277,7 +239,102 @@ public function indexNode(NodeInterface $node, $targetWorkspaceName = null)
...
+     *
+     * @param string $contextPath
+     * @param string $contextPathHash
+     * @param NodeInterface $node
+     */
+    protected function removeDuplicateDocuments($contextPath, $contextPathHash, NodeInterface $node)
+    {
+        $type = NodeTypeMappingBuilder::convertNodeTypeNameToMappingName($node->getNodeType()->getName());
+        $this->logger->log(sprintf('NodeIndexer: Check duplicate nodes for %s (%s). ContentContextHash: %s', $contextPath, $type, $contextPathHash), LOG_DEBUG, null, 'ElasticSearch (CR)');
+        $result = $this->getIndex()->request('GET', '/_search?scroll=1m&search_type=scan', [], json_encode([
```
Comment (ComiR) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: `&search_type=scan` deprecated in 2.1.0: "`scan` does not provide any benefits over a regular `scroll` request sorted by `_doc`."

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php [hunk 2]
```diff
@@ -277,7 +239,102 @@ public function indexNode(NodeInterface $node, $targetWorkspaceName = null)
...
+                        ]
+                    ]
+                ]
+            ]
+        ]));
+        $treatedContent = $result->getTreatedContent();
+        $scrollId = $treatedContent['_scroll_id'];
+
+        $result = $this->getIndex()->request('GET', '/_search/scroll?scroll=1m', [], $scrollId, false);
+        $treatedContent = $result->getTreatedContent();
```
Comment (ComiR) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: Immediately overwrites the previous `$treatedContent`, so first search result batch is never used.
Or does ES handle this differently than I think?

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php [hunk 3]
```diff
@@ -277,7 +239,102 @@ public function indexNode(NodeInterface $node, $targetWorkspaceName = null)
...
+        while (isset($treatedContent['hits']['hits']) && $treatedContent['hits']['hits'] !== []) {
+            $hits = $treatedContent['hits']['hits'];
+            $this->logger->log(sprintf('NodeIndexer: Check duplicate nodes for %s (%s), found %d document(s). ContentContextHash: %s', $contextPath, $type, count($hits), $contextPathHash), LOG_DEBUG, null, 'ElasticSearch (CR)');
+            $result = $this->getIndex()->request('GET', '/_search/scroll?scroll=1m', [], $scrollId, false);
+            $treatedContent = $result->getTreatedContent();
+        }
     }
```
Comment (ComiR) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: nothing happening here anyway...

Comment (dfeyer) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: This parts is not finished ;)

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php (hunk 2 above)
Comment (dfeyer) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: The first treatedContent contains the scroll_id, the second the "real" treatedContent

Comment (ComiR) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: Did you try that? In my case it already comes with content and the second request doesn't contain that content again...

**Comment by baschny** [comment]
@dfeyer nice work on that so far! Do you have any plan on when this will be finished / released? Do you need help, input from the outside or motivational words? We would love to be able to use ES 2.x in our projects in the near future.

**Comment by dfeyer** [comment]
@dimaip mention a problem with the cleanup command in Slack: https://neos-project.slack.com/archives/guild-search/p1473701328000019

    Catchable Fatal Error: Object of class stdClass could not be converted to string in /data/www/releases/current/Data/Temporary/Development/Cache/Code/Flow_Object_Classes/Flowpack_ElasticSearch_ContentRepositoryAdaptor_Command_NodeIndexCommandController.php line 229

And look like StyleCI is ... wrong again ...

**Comment by dimaip** [comment]
Tested on 50k nodes website: index builds, cleanup works. I love you guys!

**Comment by dfeyer** [comment]
@dimaip So push your 👍 on the first comment, and nice that this work for you. What version of ElasticSearch ?

**Comment by dfeyer** [comment]
Look like travis block on something ... i restart the test, and check tomorrow is something is broken

@ComiR Can you post your policy.java file ? Mine is currently pretty permissive ;)

**Comment by ComiR** [comment]
Mine too. See https://github.com/ElementareTeilchen/Flowpack.ElasticSearch.ContentRepositoryAdaptor/blob/task-support-es20/README.md#java-security-manager-policy-file-javapolicy

**Comment by dimaip** [comment]
@dfeyer latest ES from the official docker container

File: Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php [hunk 4]
```diff
@@ -287,6 +239,124 @@ public function indexNode(NodeInterface $node, $targetWorkspaceName = null)
...
+                            '_retry_on_conflict' => 3
+                        ]
+                    ],
+                    // http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/docs-update.html
+                    [
+                        'script' => [
+                            'inline' => '
+                                    fulltext = (ctx._source.containsKey("__fulltext") ? ctx._source.__fulltext : new HashMap());
+                                    fulltextParts = (ctx._source.containsKey("__fulltextParts") ? ctx._source.__fulltextParts : new HashMap());
+                                    _source = newData;
```
Comment (dimaip) [inline Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php]: This should be `ctx._source = newData;`

**Comment by dfeyer** [comment]
Yheaa travis is green again with ES 2.4

**Pull request merged**